### PR TITLE
chore: upgrade to Arrow 29.0 and use workspace package and dependencies

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

As PR title.

Have to use Datafusion github HEAD, because officially released 15.0 still use Arrow 28.0.

Also use new cargo feature, workspace package and dependencies, to better organized our dependencies versions.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Comment by waynexia** [comment]
We used to re-export some common deps, like using `arrow` re-exported from `datatypes`. I think this can be changed to use those deps directly. Can you check and change this?

**Comment by evenyag** [comment]
> We used to re-export some common deps, like using `arrow` re-exported from `datatypes`. I think this can be changed to use those deps directly. Can you check and change this?

I prefer to keep those re-exports so the user of `datatypes` don't have to import the arrow crate manually. This seems to be allowed by the [style guide](https://pingcap.github.io/style-guide/rust/modules.html). Arrow is part of datatypes' public API.
> Re-exporting items from dependent crates is fine if the item is logically part of the API of the re-exporting crate.

**Comment by MichaelScofield** [comment]
@waynexia I remembered Arrow is only directly used in crate datatype, other crates use Arrow indirectly, from the re-exports of datatype.

**Comment by waynexia** [comment]
Thanks for your reply @MichaelScofield @evenyag.

Another question, which dependency should be counted as "workspace dependency"? Do we have some standard for this? I prefer to treat dependencies that occur more than once as workspace dependencies.

**Comment by MichaelScofield** [comment]
@waynexia some very basic building blocks like Arrow and Datafusion; or some commonly used scaffold utils like futures and tokio

File: src/query/src/query_engine/state.rs [hunk 1]
```diff
@@ -122,32 +114,28 @@ impl QueryEngineState {
...
+            if let TableReference::Bare { table } = name {
+                TableReference::Partial { schema, table }
+            } else {
+                name
             }
-            _ => state.get_table_provider(name),
-        }
+        } else {
+            name
+        };
```
Comment (fengjiachun) [inline src/query/src/query_engine/state.rs:125]: A bit complex， how about change to:

```
let name = if let (Some(schema), TableReference::Bare { table }) = (schema, name) {
    TableReference::Partial { schema, table }
} else {
    name
};
```

**Review by fengjiachun — APPROVED** [review]

File: src/frontend/src/table/scan.rs [hunk 2]
```diff
@@ -105,9 +105,11 @@ impl DatanodeInstance {
...
         }
 
-        builder
-            .limit(0, table_scan.limit)
-            .context(error::BuildDfLogicalPlanSnafu)?;
+        if table_scan.limit.is_some() {
+            builder = builder
+                .limit(0, table_scan.limit)
+                .context(error::BuildDfLogicalPlanSnafu)?;
+        }
```
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/frontend/src/table/scan.rs:112]: "limit" is not implemented yet, tracked in #786

**Review by waynexia — APPROVED** [review]

**Pull request merged**